File: apps/reporting-service/app/domain/models/events.py
from typing import Annotated, Literal, Union

from pydantic import BaseModel, ConfigDict


class OrderCreated(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    type: Literal["order.created"]
    orderId: str
    total: float


class OrderUpdated(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    type: Literal["order.updated"]
    orderId: str
    status: Literal["PENDING", "CONFIRMED", "SHIPPED", "DELIVERED", "CANCELLED"]


OrderEvent = Annotated[Union[OrderCreated, OrderUpdated], ...]

File: packages/shared-types/src/events.ts
import type { OrderStatus } from './order'

export interface OrderCreatedEvent {
  type: 'order.created'
  orderId: string
  total: number
}

export interface OrderUpdatedEvent {
  type: 'order.updated'
  orderId: string
  status: OrderStatus
}

export type OrderEvent = OrderCreatedEvent | OrderUpdatedEvent
